woocommerce-nosto-tagging.php

```diff
@@ -645,27 +656,81 @@ public function add_page_bottom_elements() {
 		$default_element_ids = array(
 			'nosto-page-bottom',
 		);
-		$element_ids         = apply_filters( 'wcnt_add_page_bottom_elements', $default_element_ids );
-		if ( is_array( $element_ids ) && ! empty( $element_ids ) ) {
-			$this->render( 'nosto-elements', array( 'element_ids' => $element_ids ) );
+		$element_ids = apply_filters( 'wcnt_add_page_bottom_elements', $default_element_ids );
+		if ( is_array( $element_ids ) && !empty( $element_ids ) ) {
+			$this->render( 'nosto-elements', array('element_ids' => $element_ids) );
 		}
 	}
 
 	/**
+	 * Add top slots to home page
+	 *
+	 * @return array
+	 */
+	public function add_homepage_top_elements()
+	{
+		if (is_shop()) {
+			$default_element_ids = array(
+				'frontpage-nosto-1',
+			);
+			$element_ids = apply_filters('wcnt_add_page_top_elements', $default_element_ids);
+			if ( is_array( $element_ids ) && !empty( $element_ids ) ) {
+				$this->render( 'nosto-elements', array( 'element_ids' => $element_ids ), self::PAGE_TYPE_FRONT_PAGE );
+			}
+		}
+	}
+
+	/**
+	 * Add bottom slots to home page
+	 *
+	 * @return array
+	 */
+	public function add_homepage_bottom_elements()
+	{
+		if (is_shop()) {
+			$default_element_ids = array(
+				'frontpage-nosto-2',
+			);
+			$element_ids = apply_filters( 'wcnt_add_page_bottom_elements', $default_element_ids );
+			if ( is_array( $element_ids ) && !empty( $element_ids ) ) {
+				$this->render( 'nosto-elements', array( 'element_ids' => $element_ids ) );
+			}
+		}
+	}
+
+	/**
+	 * Add slots to 404 page
+	 *
+	 * @return array
+	 */
+	public function add_notfoundpage_elements()
+	{
+		$default_element_ids = array(
+			'notfound-nosto-1',
+			'notfound-nosto-2',
+			'notfound-nosto-3',
+		);
+		$this->render( 'nosto-elements', array( 'element_ids' => $default_element_ids ), self::PAGE_TYPE_NOTFOUND );
+	}
+
+	/**
 	 * Renders a template file.
 	 *
 	 * The file is expected to be located in the plugin "templates" directory.
 	 *
 	 * @since 1.0.0
 	 * @param string $template The name of the template
-	 * @param array  $data     The data to pass to the template file
+	 * @param array $data The data to pass to the template file
 	 */
-	public function render( $template, $data = array() ) {
+	public function render( $template, $data = array(), $page_type=null ) {
```

Comment:
I recommend creating an overloaded method that does this: `$this->render( 'nosto-elements', array( 'element_ids' => $default_element_ids ), self::PAGE_TYPE_FRONT_PAGE );` It should take the element ids and the page type as the param. How does this sound?